Allow disable comment to be the directly above the statement

pndewit opened the pull request:
The comment line to disable the postcss-lit plugin (`postcss-lit-disable-next-line`) doesn't necessarily need to be on the parent statement, especially with static class fields being more and more mainstream: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static

The following snippet breaks stylelint with postcss-lit:

```javascript
const MY_CONST = 600;

export class MyClass extends LitElement {
  static styles = css`
    @media (width >= ${MY_CONST}px) {
      ...
    }
  `;
}
```

Errors:
- Unexpected invalid media query        media-query-no-invalid
   "(width >= POSTCSS_LIT_0px)"
- Unexpected unknown media feature      media-feature-name-no-unknown
   name "POSTCSS_LIT_0px"

To disable the plugin, I'd now have to disable it for the entire class because `styles` is a static class field and therefore the parent statement is the `MyClass` class:
```javascript
const MY_CONST = 600;

// postcss-lit-disable-next-line
export class MyClass extends LitElement {
  static styles = css`
    @media (width >= ${MY_CONST}px) {
      ...
    }
  `;
}
```

Because my classes contain more CSS than just this media statement, I'd like to be able to not disable this plugin for the entire class nor do I want to replace my static class field back to a getter. This PR allows disable comments to be placed directly above the statement, but the previous behaviour of looking at the parent comments is taking precedence so that it should not be a breaking change. This now works:

```javascript
const MY_CONST = 600;

export class MyClass extends LitElement {
  static styles = [
    // postcss-lit-disable-next-line
    css`
      @media (width >= ${MY_CONST}px) {
        ...
      }
    `,
    css`
      .my-class-i-want-to-stylelint {
        ...
      }
    `,
  ];
}
```

I also made it so the disable comment doesn't necessarily need to be the last comment before the statement. It can be one of the comments which is required if I also want to disable for example eslint for the same line, e.g.:

```javascript
// postcss-lit-disable-next-line
// eslint-disable-next-line
css`
  @media (width >= ${MY_CONST}px) {
    ...
  }
`
```

Let me know what you think, I'm open for suggestions!

---

EDIT:
Future work: maybe think about also supporting the following syntax:

```javascript
css`
  /* postcss-lit-disable-next-line */
  .foo { color: hotpink; }
`;
```

As far as I can see the CSS tagged template expression is not parsed by Babel though, so it might require using another library for that.

43081j commented on src/util.ts:
```diff
@@ -8,7 +8,7 @@ import {TaggedTemplateExpression, Comment} from '@babel/types';
 export function isDisableComment(node: Comment): boolean {
   return (
-    node.type === 'CommentLine' &&
+    node && node.type === 'CommentLine' &&
```
this will always be truthy since `node` isn't optional (and the `some` further down will always pass a node)

so you can probably revert this to what it was before

pndewit replied:
Makes sense, solved!

43081j commented:
seems to make sense to me

I think it was done originally because we may be looking at a deeper node and wanted to look for comments on the owning statement instead

looks like we're still doing that with this change so should be fine

pndewit commented:
Yep, tried to keep the current logic as the first check so it wouldn't introduce breaking changes :)

43081j commented:
could you add some tests for this too?

just also wondered - won't there always be a statement parent somewhere, so it'll never be nullish?

tests will help prove it out though

pndewit commented:
I've tried to add a test case, but it's quite hard as the AST will get more than a simple css expression as all the other tests are using.

The old code was handling the case that getStatementParent returns something nullish:

```javascript
const statement = path.getStatementParent();

if (statement && statement.node.leadingComments) { ... }
```

So I left that handling intact by adding optional chaining.

pndewit commented:
I now see I made a boo-boo, my change only enables this to work (note the array):

```javascript
const MY_CONST = 600;

class MyClass {
  static styles = [
    // postcss-lit-disable-next-line
    css`
      @media (width >= ${MY_CONST}px) {
        .foo { color: hotpink; }
      }
    `,
  ];
}
```

But I also need this to work (note the absence of the array):

```javascript
const MY_CONST = 600;

class MyClass {
  // postcss-lit-disable-next-line
  static styles = css`
    @media (width >= ${MY_CONST}px) {
      .foo { color: hotpink; }
    }
  `;
}
```

pndewit commented:
> I now see I made a boo-boo, my change only enables this to work (note the array):
> 
> ```js
> const MY_CONST = 600;
> 
> class MyClass {
>   static styles = [
>     // postcss-lit-disable-next-line
>     css`
>       @media (width >= ${MY_CONST}px) {
>         .foo { color: hotpink; }
>       }
>     `,
>   ];
> }
> ```
> 
> But I also need this to work (note the absence of the array):
> 
> ```js
> const MY_CONST = 600;
> 
> class MyClass {
>   // postcss-lit-disable-next-line
>   static styles = css`
>     @media (width >= ${MY_CONST}px) {
>       .foo { color: hotpink; }
>     }
>   `;
> }
> ```

I fixed this by also checking the node parent.

This also fixed the tests I tried to add:
1. should ignore a child node -> This one wasn't working before my latest pushes
2. should ignore a child node from an array -> This one was working before my latest pushes and continues to work
3. should ignore everything inside a disabled parent node -> This one was always working, even without my changes

43081j commented:
It's a little too specific checking the parent and the current node I think

I had a quick dig and I think what we're looking for is `findParent`:

```ts
let parentStatement = undefined;
const disableNode = path.findParent((p) => {
  // Determine if this is the statement
  if (!parentStatement) {
    if (!p.parentPath || (Array.isArray(p.container) && p.isStatement()) {
      parentStatement = p;
    }
  } else {
    // We already saw the root statement, stop looking
    return false; 
  }
  // Check if this parent has a disable comment 
  if (hasDisableComment(p)) {
    return true; 
  }
  return false; 
});
if (disableNode) {
  // It is disabled
}
```

So this will support any parent node having a disabled comment until we reach the root statement

I wrote this on mobile so some names are probably wrong

pndewit commented:
How about this? :)
